File: banchus-app/src/main/java/pe/nanamochi/banchus/controller/resource/AvatarController.java
package pe.nanamochi.banchus.controller.resource;

import lombok.RequiredArgsConstructor;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import pe.nanamochi.banchus.service.StorageService;

@RestController
@RequestMapping("/")
@RequiredArgsConstructor
public class AvatarController {
  private final StorageService storageService;

  @GetMapping(value = "/{userId}", produces = MediaType.IMAGE_PNG_VALUE)
  public ResponseEntity<byte[]> getAvatar(@PathVariable String userId) {
    byte[] avatarData = storageService.getAvatar(userId);
    return ResponseEntity.ok().contentType(MediaType.IMAGE_PNG).body(avatarData);
  }
}

File: banchus-app/src/main/kotlin/pe/nanamochi/banchus/controller/client/resource/AvatarController.kt
package pe.nanamochi.banchus.controller.client.resource

import com.github.michaelbull.result.getOrThrow
import com.github.michaelbull.result.mapError
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import org.springframework.web.server.ResponseStatusException
import pe.nanamochi.banchus.service.StorageService

@RestController
@RequestMapping("/")
class AvatarController(private val storageService: StorageService) {
    @GetMapping("/{userId}", produces = [MediaType.IMAGE_PNG_VALUE])
    fun getAvatar(@PathVariable userId: String): ResponseEntity<ByteArray> {
        val bytes =
            storageService
                .getAvatar(userId)
                .mapError { ResponseStatusException(HttpStatus.NOT_FOUND) }
                .getOrThrow { it }

        return ResponseEntity.ok().body(bytes)
    }
}
